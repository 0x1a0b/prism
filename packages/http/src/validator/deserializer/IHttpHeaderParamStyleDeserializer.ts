export type DeserializeHttpHeader = (value: string, type: string, explode: boolean) => any;

export interface IHttpHeaderParamStyleDeserializer {
  supports: (style: string) => boolean;
  deserialize: DeserializeHttpHeader;
}
