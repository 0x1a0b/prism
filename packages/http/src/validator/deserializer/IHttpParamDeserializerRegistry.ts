export interface IHttpParamDeserializerRegistry<T extends Function> {
  get(style: string): T | undefined;
}
