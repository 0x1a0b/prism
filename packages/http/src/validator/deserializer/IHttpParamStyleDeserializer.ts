export interface IHttpParamStyleDeserializer<T> {
  supports: (style: string) => boolean;
  deserialize: T;
}
