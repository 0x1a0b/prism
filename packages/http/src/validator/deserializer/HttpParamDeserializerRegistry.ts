import { IHttpParamStyleDeserializer } from '@stoplight/prism-http/validator/deserializer/IHttpParamStyleDeserializer';

export class HttpParamDeserializerRegistry<
  D extends IHttpParamStyleDeserializer<T>,
  T extends Function
> {
  constructor(private deserializers: D[]) {}

  public get(style: string): T | undefined {
    const deserializer = this.deserializers.find(d => d.supports(style));

    if (!deserializer) {
      return;
    }

    return deserializer.deserialize.bind(deserializer);
  }
}
