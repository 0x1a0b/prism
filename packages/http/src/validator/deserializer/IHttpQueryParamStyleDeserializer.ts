import { ISchema } from '@stoplight/types/schema';

import { IHttpParamStyleDeserializer } from './IHttpParamStyleDeserializer';

export type DeserializeHttpQuery = (
  key: string,
  query: {
    [name: string]: string | string[];
  },
  schema: ISchema,
  explode: boolean
) => any;

export interface IHttpQueryParamStyleDeserializer
  extends IHttpParamStyleDeserializer<DeserializeHttpQuery> {
  deserialize: DeserializeHttpQuery;
}
