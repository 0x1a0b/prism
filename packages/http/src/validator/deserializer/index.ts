import { HttpParamDeserializerRegistry } from './HttpParamDeserializerRegistry';
import { DeepObjectStyleDeserializer } from './style/DeepObjectStyleDeserializer';
import { DelimitedStyleDeserializer } from './style/DelimitedStyleDeserializer';
import { FormStyleDeserializer } from './style/FormStyleDeserializer';
import { SimpleStyleDeserializer } from './style/SimpleStyleDeserializer';
import {
  DeserializeHttpHeader,
  DeserializeHttpQuery,
  IHttpHeaderParamStyleDeserializer,
  IHttpQueryParamStyleDeserializer,
} from './types';

export const header = new HttpParamDeserializerRegistry<
  IHttpHeaderParamStyleDeserializer,
  DeserializeHttpHeader
>([new SimpleStyleDeserializer()]);

export const query = new HttpParamDeserializerRegistry<
  IHttpQueryParamStyleDeserializer,
  DeserializeHttpQuery
>([
  new FormStyleDeserializer(),
  new DelimitedStyleDeserializer('%20', 'spaceDelimited'),
  new DelimitedStyleDeserializer('|', 'pipeDelimited'),
  new DeepObjectStyleDeserializer(),
]);
