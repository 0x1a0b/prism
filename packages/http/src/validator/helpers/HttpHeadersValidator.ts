import { IValidation } from '@stoplight/prism-core';
import { ValidationSeverity } from '@stoplight/prism-core/types';
import { HttpParamDeserializerRegistry } from '@stoplight/prism-http/validator/deserializer/HttpParamDeserializerRegistry';
import {
  DeserializeHttpHeader,
  IHttpHeaderParamStyleDeserializer,
} from '@stoplight/prism-http/validator/deserializer/IHttpHeaderParamStyleDeserializer';
import { convertAjvErrors } from '@stoplight/prism-http/validator/helpers/convertAjvErrors';
import { IHttpContent, IHttpRequest } from '@stoplight/types/http';
import * as Ajv from 'ajv';
import { Ajv as AjvClass } from 'ajv';

export class HttpHeadersValidator {
  private ajv: AjvClass;

  constructor(
    private readonly registry: HttpParamDeserializerRegistry<
      IHttpHeaderParamStyleDeserializer,
      DeserializeHttpHeader
    >
  ) {
    this.ajv = new Ajv({ allErrors: true, messages: true });
  }

  public validate(
    headers: { [name: string]: string } = {},
    requestSpec?: IHttpRequest,
    mediaType?: string
  ) {
    if (!requestSpec) {
      return [];
    }

    if (!requestSpec.headers) {
      return [];
    }

    return requestSpec.headers.reduce<IValidation[]>((results, spec) => {
      if (!headers.hasOwnProperty(spec.name) && spec.required === true) {
        results.push({
          path: ['header', spec.name],
          name: 'required',
          summary: '',
          message: `Missing ${spec.name} header param`,
          severity: ValidationSeverity.ERROR,
        });

        // stop further checks
        return results;
      }

      const content = this.resolveContent(spec.content, mediaType);

      if (content && content.schema) {
        const deserialize = this.registry.get(spec.style || 'simple');

        if (deserialize) {
          Array.prototype.push.apply(
            results,
            this.validateAgainstSchema(
              deserialize(headers[spec.name], content.schema.type, spec.explode || false),
              content.schema
            )
          );
        }
      }

      if (spec.deprecated === true) {
        results.push({
          path: ['header', spec.name],
          name: 'deprecated',
          summary: '',
          message: `Header param ${spec.name} is deprecated`,
          severity: ValidationSeverity.WARN,
        });
      }

      return results;
    }, []);
  }

  private validateAgainstSchema(value: string, schema: any): IValidation[] {
    const validate = this.ajv.compile(schema);

    if (!validate(value)) {
      return convertAjvErrors(validate.errors, ValidationSeverity.ERROR).map(error =>
        Object.assign({}, error, { path: ['header', ...error.path] })
      );
    }

    return [];
  }

  private resolveContent(content: { [mediaType: string]: IHttpContent }, mediaType?: string) {
    if (!mediaType) {
      if (content.hasOwnProperty('*')) {
        return content['*'];
      }

      return null;
    }

    if (content.hasOwnProperty(mediaType)) {
      return content[mediaType];
    }

    return content['*'];
  }
}
