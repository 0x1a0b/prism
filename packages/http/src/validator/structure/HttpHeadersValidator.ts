import { IValidation, ValidationSeverity } from '@stoplight/prism-core';
import { IHttpHeaderParam } from '@stoplight/types/http';

import { DeserializeHttpHeader, IHttpParamDeserializerRegistry } from '../deserializer/types';
import { resolveContent } from '../helpers/resolveContent';
import { validateAgainstSchema } from '../helpers/validateAgainstSchema';
import { IHttpHeadersValidator } from './IHttpHeadersValidator';

export class HttpHeadersValidator implements IHttpHeadersValidator {
  constructor(private readonly registry: IHttpParamDeserializerRegistry<DeserializeHttpHeader>) {}

  public validate(
    headers: { [name: string]: string } = {},
    headerSpecs: IHttpHeaderParam[],
    mediaType?: string
  ): IValidation[] {
    return headerSpecs.reduce<IValidation[]>((results, spec) => {
      if (!headers.hasOwnProperty(spec.name) && spec.required === true) {
        results.push({
          path: ['header', spec.name],
          name: 'required',
          summary: '',
          message: `Missing ${spec.name} header param`,
          severity: ValidationSeverity.ERROR,
        });

        // stop further checks
        return results;
      }

      const content = resolveContent(spec.content, mediaType);

      if (content && content.schema) {
        const deserialize = this.registry.get(spec.style || 'simple');

        if (deserialize) {
          Array.prototype.push.apply(
            results,
            validateAgainstSchema(
              deserialize(headers[spec.name], content.schema.type, spec.explode || false),
              content.schema,
              'header'
            )
          );
        }
      }

      if (spec.deprecated === true) {
        results.push({
          path: ['header', spec.name],
          name: 'deprecated',
          summary: '',
          message: `Header param ${spec.name} is deprecated`,
          severity: ValidationSeverity.WARN,
        });
      }

      return results;
    }, []);
  }
}
