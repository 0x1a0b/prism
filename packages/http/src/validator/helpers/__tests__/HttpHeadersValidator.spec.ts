import { IHttpParamDeserializerRegistry } from '@stoplight/prism-http/validator/deserializer/IHttpParamDeserializerRegistry';
import { DeserializeHttpHeader } from '../../deserializer/IHttpHeaderParamStyleDeserializer';
import { HttpHeadersValidator } from '../HttpHeadersValidator';

describe('HttpHeadersValidator', () => {
  const httpParamDeserializerRegistry = {
    deserializers: [],
    get: () => v => v,
  } as IHttpParamDeserializerRegistry<DeserializeHttpHeader>;
  const httpHeadersValidator = new HttpHeadersValidator(httpParamDeserializerRegistry);

  beforeEach(() => {
    jest.clearAllMocks();
    jest.spyOn(httpParamDeserializerRegistry, 'get');
  });

  describe('validate()', () => {
    describe('headers are missing', () => {
      it('returns empty validation error list', () => {
        expect(
          httpHeadersValidator.validate(undefined, { headers: [] }, 'application/json')
        ).toEqual([]);
      });
    });

    describe('spec for request is missing', () => {
      it('returns empty validation error list', () => {
        expect(httpHeadersValidator.validate({}, undefined, 'application/json')).toEqual([]);
      });
    });

    describe('spec for headers is missing', () => {
      it('returns empty validation error list', () => {
        expect(httpHeadersValidator.validate({}, {}, 'application/json')).toEqual([]);
      });
    });

    describe('spec is present', () => {
      describe('header is not present', () => {
        describe('spec defines it as required', () => {
          it('returns validation error', () => {
            expect(
              httpHeadersValidator.validate(
                {},
                { headers: [{ name: 'aHeader', required: true, content: {} }] }
              )
            ).toMatchSnapshot();
          });
        });
      });

      describe('header is present', () => {
        describe('schema is present', () => {
          describe('deserializer not available', () => {
            it('omits schema validation', () => {
              jest.spyOn(httpParamDeserializerRegistry, 'get').mockReturnValueOnce(undefined);

              expect(
                httpHeadersValidator.validate(
                  { 'x-test-header': 'abc' },
                  {
                    headers: [
                      {
                        name: 'x-test-header',
                        content: { '*': { mediaType: '*', schema: { type: 'number' } } },
                      },
                    ],
                  }
                )
              ).toEqual([]);
            });
          });

          describe('deserializer is available', () => {
            describe('header is valid', () => {
              it('validates positively against schema', () => {
                expect(
                  httpHeadersValidator.validate(
                    { 'x-test-header': 'abc' },
                    {
                      headers: [
                        {
                          name: 'x-test-header',
                          content: { '*': { mediaType: '*', schema: { type: 'string' } } },
                        },
                      ],
                    }
                  )
                ).toEqual([]);
              });
            });

            describe('header is not valid', () => {
              it('not validates against schema', () => {
                expect(
                  httpHeadersValidator.validate(
                    { 'x-test-header': 'abc' },
                    {
                      headers: [
                        {
                          name: 'x-test-header',
                          content: { '*': { mediaType: '*', schema: { type: 'number' } } },
                        },
                      ],
                    }
                  )
                ).toMatchSnapshot();
              });
            });
          });
        });

        describe('mediaType is provided', () => {
          describe('specific mediaType found in spec content list', () => {
            it('validates against more specific spec', () => {
              expect(
                httpHeadersValidator.validate(
                  { 'x-test-header': 'abc' },
                  {
                    headers: [
                      {
                        name: 'x-test-header',
                        content: {
                          '*': { mediaType: '*', schema: { type: 'string' } },
                          'application/testson': {
                            mediaType: 'application/testson',
                            schema: { type: 'number' },
                          },
                        },
                      },
                    ],
                  },
                  'application/testson'
                )
              ).toMatchSnapshot();
            });
          });

          describe('specific mediaType not found in spec content list', () => {
            it('validates against default spec', () => {
              expect(
                httpHeadersValidator.validate(
                  { 'x-test-header': 'abc' },
                  {
                    headers: [
                      {
                        name: 'x-test-header',
                        content: {
                          '*': { mediaType: '*', schema: { type: 'string' } },
                          'application/testson': {
                            mediaType: 'application/testson',
                            schema: { type: 'number' },
                          },
                        },
                      },
                    ],
                  },
                  'application/failson'
                )
              ).toMatchSnapshot();
            });
          });
        });

        describe('deprecated flag is set', () => {
          it('returns deprecation warning', () => {
            expect(
              httpHeadersValidator.validate(
                { 'x-test-header': 'abc' },
                {
                  headers: [
                    {
                      name: 'x-test-header',
                      deprecated: true,
                      content: {},
                    },
                  ],
                }
              )
            ).toMatchSnapshot();
          });
        });
      });
    });
  });
});
