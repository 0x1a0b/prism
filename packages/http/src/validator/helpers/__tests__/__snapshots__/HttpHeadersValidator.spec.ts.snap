// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HttpHeadersValidator validate() spec is present header is not present spec defines it as required returns validation error 1`] = `
Array [
  Object {
    "message": "Missing aHeader header param",
    "name": "required",
    "path": Array [
      "header",
      "aHeader",
    ],
    "severity": "error",
    "summary": "",
  },
]
`;

exports[`HttpHeadersValidator validate() spec is present header is present deprecated flag is set returns deprecation warning 1`] = `
Array [
  Object {
    "message": "Header param x-test-header is deprecated",
    "name": "deprecated",
    "path": Array [
      "header",
      "x-test-header",
    ],
    "severity": "warn",
    "summary": "",
  },
]
`;

exports[`HttpHeadersValidator validate() spec is present header is present mediaType is provided specific mediaType found in spec content list validates against more specific spec 1`] = `
Array [
  Object {
    "message": "should be number",
    "name": "type",
    "path": Array [
      "header",
    ],
    "severity": "error",
    "summary": "should be number",
  },
]
`;

exports[`HttpHeadersValidator validate() spec is present header is present mediaType is provided specific mediaType not found in spec content list validates against default spec 1`] = `Array []`;

exports[`HttpHeadersValidator validate() spec is present header is present schema is present deserializer is available header is not valid not validates against schema 1`] = `
Array [
  Object {
    "message": "should be number",
    "name": "type",
    "path": Array [
      "header",
    ],
    "severity": "error",
    "summary": "should be number",
  },
]
`;
